include/ck_tile/ops/epilogue/cshuffle_epilogue.hpp

```diff
@@ -1,194 +1,194 @@
+    }
 
-        // Move the window
-        move_tile_window(o_dram_window_tmp, step);
+    CK_TILE_HOST_DEVICE static constexpr index_t GetSmemSize()
+    {
+        return kMWave * kNWave * kMPerXdl * kNPerXdl * sizeof(ODataType);
+    }
 
-        // Permute the data within the tile if necessary
-        if constexpr(kTilePermute)
-        {
-            permute_tile_data(o_acc_tile);
-        }
+    template <typename ODramWindow,
+              typename OAccTile,
+              memory_operation_enum out_memory_data_op = memory_operation_enum::set>
+    CK_TILE_DEVICE auto
+    operator()(ODramWindow& out_dram_window, const OAccTile& o_acc_tile, void* p_smem)
+    {
 
-        // Store the tile data to the permuted location
-        if constexpr(kPadM || kPadN)
-        {
+        const index_t iMWarp = get_warp_id() / kNWave;
+        const index_t iNWarp = get_warp_id() - iMWarp * kNWave;
+
+        constexpr auto lds_block_desc = MakeLdsBlockDescriptor<Problem>();
+        auto o_lds_block              = make_tensor_view<address_space_enum::lds>(
+            static_cast<ODataType*>(p_smem), lds_block_desc);
+        auto in_lds_window =
+            make_tile_window(o_lds_block,
+                             make_tuple(number<kMPerXdl>{}, number<kNPerXdl>{}),
+                             {number<kMPerXdl>{} * iMWarp, number<kNPerXdl>{} * iNWarp});
+        auto out_lds_window =
+            make_tile_window(o_lds_block,
+                             make_tuple(number<kMWave * kMPerXdl>{}, number<kNWave * kNPerXdl>{}),
+                             {0, 0});
+
+        using SFC                    = space_filling_curve<sequence<kMPerBlock, kNPerBlock>,
+                                        sequence<0, 1>,
+                                        sequence<kMPerXdl * kMWave, kNPerXdl * kNWave>>;
+        constexpr index_t num_access = SFC::get_num_of_access();
+
+        using TileEncodingPattern =
+            TileDistributionEncodingPattern2D<kBlockSize,
+                                              kMPerIteration,
+                                              kNPerIteration,
+                                              GetVectorSizeC(),
+                                              tile_distribution_pattern::thread_raked>;
+        constexpr auto dram_tile_distribution = TileEncodingPattern::Make2DStaticTileDistribution();
+
+        constexpr auto c_warp_y_lengths =
+            to_sequence(CWarpDstr{}.get_ys_to_d_descriptor().get_lengths());
+        constexpr auto c_warp_y_index_zeros = uniform_sequence_gen_t<CWarpDstr::NDimY, 0>{};
+
+        CWarpTensor c_warp_in_tensor;
+        static_for<0, num_access, 1>{}([&](auto iAccess) {
+            constexpr auto idx_y_start = SFC::get_index(iAccess);
+
+            constexpr auto mIter = number<idx_y_start.at(number<0>{}) / (kMPerXdl * kMWave)>{};
+            constexpr auto nIter = number<idx_y_start.at(number<1>{}) / (kNPerXdl * kNWave)>{};
```

Comment:
I wonder whether it wouldn't be more optimal to kind of turn around this relation. I mean SFC could iterate over `mIter` and `nIter` and then you would calculate offsets. This way you would do multiplication instead of division which I think is more expensive.

Reply:
It is constexpr calculation, so it doesnt has an impact on perf